### PR TITLE
Delete userprofile

### PR DESCRIPTION
**Pull request by KOTLAPRA23**
### Change description ###

feat: Delete user profile

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

**Comment by shashank-rastogi** [comment]
@KOTLAPRA23  - how r u going to toggle this in production?

File: src/main/resources/db/migration/V1_5__adding_on_delete_cascade_tables.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+ALTER TABLE response DROP CONSTRAINT user_profile_id_fk1;
+alter table response add constraint user_profile_id_fk1 foreign key (user_profile_id)references user_profile (id) on delete cascade;
```
Comment (shashank-rastogi) [inline src/main/resources/db/migration/V1_5__adding_on_delete_cascade_tables.sql:3]: response table should not have delete cascade as you will never be deleting it from response table

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        List<UserProfile> userProfiles = new ArrayList<UserProfile>();
+        Set<String> userIds = new HashSet<String>(profilesData.getUserIds());
+        userIds.forEach(userId -> userProfiles.add(validateUserStatus(userId)));
+
+        return deleteUserProfiles(userProfiles);
+    }
+
+    /**
+     * Either delete all the audit and userProfiles data from the data base or none.
+     *
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:45]: can u change the comment here?.. i dont think u r deleting anything from the response table

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    private UserProfilesDeletionResponse deleteUserProfiles(List<UserProfile> userProfiles) {
+
+        userProfileRepository.deleteAll(userProfiles);
+        UserProfilesDeletionResponse deletionResponse = new UserProfilesDeletionResponse();
+        HttpStatus status = HttpStatus.NO_CONTENT;
+        deletionResponse.setMessage("UserProfiles Successfully Deleted");
+        deletionResponse.setStatusCode(status.value());
+        auditService.persistAudit(deletionResponse);
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:54]: it should persist the user id deleted from the profile table

File: src/test/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Test(expected = ResourceNotFoundException.class)
+    public void testShouldThrowExceptionWhenEmptyUserProfileToDelete() {
+
+        List<String> userIds = new ArrayList<String>();
+        userIds.add("1234");
+        UserProfileDataRequest userProfilesDeletionData = new UserProfileDataRequest(userIds);
+        when(userProfileRepositoryMock.findByIdamId(any(String.class))).thenReturn(Optional.ofNullable(any()));
+
```
[the remark this answers is not in the document]
Comment (KOTLAPRA23) [reply src/test/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImplTest.java:82]: currently throwing 404(ResourceNotFoundException) only if user profile not present in the DB.

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java (hunk 2 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:45]: I am deleting the entries from response and user profile table.

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java (hunk 3 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:54]: not at the moment.

File: src/main/resources/db/migration/V1_5__adding_on_delete_cascade_tables.sql (hunk 1 above)
Comment (KOTLAPRA23) [reply src/main/resources/db/migration/V1_5__adding_on_delete_cascade_tables.sql:3]: currently we are deleting entries from both user_profile and response table and delete cascade adding on the user_profile_id_fk1 constraint

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java (hunk 3 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java]: now user id will store in the audit table

File: src/main/resources/db/migration/V1_5__adding_on_delete_cascade_tables.sql (hunk 1 above)
Comment (KOTLAPRA23) [reply src/main/resources/db/migration/V1_5__adding_on_delete_cascade_tables.sql:3]: https://tools.hmcts.net/jira/browse/RDCC-1805 i have created tech debt ticket for remove the foreign key constraint and modify the code

**Pull request merged**